## More about Algorithms

Let's return to the idea of an algorithm.  We already mentioned that an algorithm is the abstract set of steps a program will execute, as separate from the specific language it is written in.  In another, slightly different, sense an algorithm is a plan for solving a specific problem.  Computer scientists study all kinds of problems.  Some of these are big problems with clear applications to data science.  For example, how do you approximate the solution to a linear regression for a large dataset? Other problems are more foundational in nature.  Here, let's consider the problem of how to find the square root of a positive number, $x$.  Of course, in Python, you can just type in $x$^$0.5$, but let's pretend we don't know that (or we can pretend that we're trying to write an interpreter that knows how to take square roots.)

Previously, we used an algorithm to check if an integer is a perfect square.  We started with a guess of zero and increased it by one until the square of our guess was as large as the input.  One idea is to adapt this algorithm to find the square root of any number.  Of course, that algorithm only searched through the integers, so we'll have to take smaller steps if we want to find the square root of any number.

It's worth noting that we can never find the square root of a number exactly.  That's because square roots usually don't terminate - it would take an infinite number of digits to write them down.  What we can do instead is choose a small epsilon (say, 0.00001), and try to find a number that's less than epsilon away from the square root of $x$.  To refactor our algorithm, we'll start with a guess of zero and add epsilon in each step.  This will guarrantee that the answer we get is within epsilon of the actual square root.

Try finding the square root of 10 with this algorithm.

In [5]:
## Exhuastive Search to Find a Square Root

x = float(input("enter a number:"))
epsilon = 0.00001
num_guesses = 0
ans = 0.0
while ans*ans<= x:
    ans += epsilon
    num_guesses += 1
print('number of guesses =', num_guesses)
print(ans, 'is close to square root of', x)

enter a number:12345
number of guesses = 11110806
111.1080600240677 is close to square root of 12345.0


You can check that the answer is indeed close to the square root of 10.  Now try the program on a large number, say 12345.  How long does it take?

This is an example of a *brute force algorithm*.  It searches through all possible solutions, until the right one is found.  You can see that the brute force algorithm will always find the answer, but it can take a rather long time.

Depite this, there are many times in programming when you will want to use a brute force algorithm.  It could be that you know the number of possible solutions is small, so the brute force algorithm will work in a reasonable time.  Brute force algorithms are also very simple.  Finally, it could just be that the problem space is unstructured and there's no better choice out there.

Here's another algorithm that's well-known among computer scientists: a *bisection search*.  Instead of making a single guess, imagine a window, or an interval of numbers that we know contains the square root of $x$.  At the beginning, we can use the window from 0 to x.  Clearly, the square root of $x$ has to be in there somewhere.

Next, let's look at the midpoint of our window, which will be $x/2$ at the moment.  We can check if this point is too low or two high by squaring it and comparing it to $x$.  If the midpoint of our window is too low, we can throw away the bottom half of the window and focus on the top half.  Otherwise, we throw out the top half of the window and focus on the bottom half.  Now we have a new window, which is half as large, and we can repeat the same process.  Eventually our window will be small enough that we know the midpoint is within epsillon of the right answer.

The python script below implements a bisection search.  Notice that we use variables low and high to keep track of the bottom and the top of our window.  When the size of the window shrinks below 2 * epsillon, we know that any point inside the window is within epsillon of the midpoint.  This means that the midpoint is a valid guess for the squareroot of $x$, up to the precision we want.

In [10]:
## Bisection Search to Find a Square Root

x = float(input("enter a number:"))
epsilon = 0.00001
num_guesses = 0
low = 0.0
high = x
ans = (high + low)/2.0
while high-low >= 2 * epsilon:
    print("low =",low,"high =", high)
    num_guesses += 1
    if ans**2 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
print('number of guesses =', num_guesses)
print(ans, 'is close to square root of', x)

enter a number:.25
low = 0.0 high = 0.25
low = 0.125 high = 0.25
low = 0.1875 high = 0.25
low = 0.21875 high = 0.25
low = 0.234375 high = 0.25
low = 0.2421875 high = 0.25
low = 0.24609375 high = 0.25
low = 0.248046875 high = 0.25
low = 0.2490234375 high = 0.25
low = 0.24951171875 high = 0.25
low = 0.249755859375 high = 0.25
low = 0.2498779296875 high = 0.25
low = 0.24993896484375 high = 0.25
low = 0.249969482421875 high = 0.25
number of guesses = 14
0.24999237060546875 is close to square root of 0.25


Try running the algorithm to find the square root of 10, and of 12345.  How many steps did the algorithm take?

Now try running the algorithm to find the square root of 0.25.  What goes wrong?  Try to fix the algorithm so that it works correctly.

Finally, let's examine a third algorithm for finding square roots, known as the Babylonian method, or as Heron's method, after an ancient Greek mathematician.  This algorithm is a special case of Newton's method, invented centuries later to approximate the roots of functions.  In this case, we want to find $ans$, such that the function $f(ans) = ans^2 - x$ equals 0.

Newton's method is easiest to understand with a picture.  We begin with a guess for ans and compute our function $f$, and the derivative of $f$ at the guess.  Even though our function is a curve, we approximate it as the tangent line at our guess, and compute where the line crosses the zero.  This becomes our next guess.  Using some calculus, we can compute the next guess for ans as $(x/ans + ans)/2$.  This is encoded in the algorithm, below.

In [11]:
## Newton's Method to Find a Square Root (Heron's algorithm)
x = float(input("enter a number:"))
epsilon = 0.00001
num_guesses = 0
ans = 1
while abs(x/ans - ans)>epsilon:
    ans = (x/ans + ans)/2
    num_guesses += 1
print("number of guesses =", num_guesses)
print(ans, "is close to the square root of", x)

enter a number:12345
number of guesses = 10
111.10805770848404 is close to the square root of 12345.0


Try running this algorithm.  How does it perform compared to the exhuastive search and the bisection search algorithms?  Notice that this algorithm actually tries to use features of the problem to find a solution - like the fact that $x^2$ is a smooth function of x.